cli/ssh.go

```diff
@@ -100,17 +100,62 @@ func (r *RootCmd) ssh() *clibase.Cmd {
 			stopPolling := tryPollWorkspaceAutostop(ctx, client, workspace)
 			defer stopPolling()
 
+			// Enure connection is closed if the context is canceled or
+			// the workspace reaches the stopped state.
+			//
+			// Watching the stopped state is a work-around for cases
+			// where the agent is not gracefully shut down and the
+			// connection is left open. If, for instance, the networking
+			// is stopped before the agent is shut down, the disconnect
+			// will usually not propagate.
+			//
+			// See: https://github.com/coder/coder/issues/6180
+			wsWatch, err := client.WatchWorkspace(ctx, workspace.ID)
+			if err != nil {
+				return err
+			}
+			watchAndClose := func(c io.Closer) {
+				// Ensure session is ended on both context cancellation
+				// and workspace stop.
+				defer c.Close()
+
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case w, ok := <-wsWatch:
+						if !ok {
+							return
+						}
+
+						// Note, we only react to the stopped state here because we
+						// want to give the agent a chance to gracefully shut down
+						// during "stopping".
+						if w.LatestBuild.Status == codersdk.WorkspaceStatusStopped {
+							_, _ = fmt.Fprintf(inv.Stderr, "Workspace %q has stopped. Closing connection.\r\n", workspace.Name)
```

Comment:
Yay or nay? This seems to be visible even non-verbose SSH is used (e.g. both `ssh` and `ssh -vvv`).